File: laba4 (2)/task1.cpp
#include <iostream>
#include <cmath>
#include <vector>
#include <iomanip>

using namespace std;

// Исходная функция f(x)
double f(double x) {
    return cos(x + 0.5) - x - 2;
}

// Производная f(x) для метода Ньютона
double df(double x) {
    return -sin(x + 0.5) - 1;
}

// Функция x = phi(x) для метода итераций
double phi(double x) {
    return cos(x + 0.5) - 2;
}

int main() {
    double eps = 0.0001;
    cout << fixed << setprecision(5);

    
    cout << "--- Метод половинного деления ---" << endl;
    vector<double> aVec, bVec, diff_vec;
    double a = -2.0, b = -1.0;

    while (abs(b - a) > eps) {
        aVec.push_back(a);
        bVec.push_back(b);
        diff_vec.push_back(abs(b - a));
        
        double c = (a + b) / 2.0;
        if (f(a) * f(c) < 0) b = c;
        else a = c;
    }

    cout << "N     an          bn          bn-an" << endl;
    for (int i = 0; i < aVec.size(); i++) {
        cout << i << "     " << aVec[i] << "    " << bVec[i] << "    " << diff_vec[i] << endl;
    }
    cout << "Корень: " << (a + b) / 2.0 << endl << endl;


    
    cout << "--- Метод Ньютона ---" << endl;
    vector<double> xn_newton, xnext_newton, diff_newton;
    double xNewt = -1.5;

    for (int i = 0; i < 100; i++) {
        double x_next = xNewt - f(xNewt) / df(xNewt);
        double d = abs(x_next - xNewt);
        
        xn_newton.push_back(xNewt);
        xnext_newton.push_back(x_next);
        diff_newton.push_back(d);

        xNewt = x_next;
        if (d < eps) break;
    }

    cout << "N     xn          xn+1        |xn+1-xn|" << endl;
    for (int i = 0; i < xn_newton.size(); i++) {
        cout << i << "     " << xn_newton[i] << "    " << xnext_newton[i] << "    " << diff_newton[i] << endl;
    }
    cout << "Корень: " << xNewt << endl << endl;


   
    cout << "--- Метод простых итераций ---" << endl;
    vector<double> xnIter, xnext_iter, diffIter;
    double xIt = -1.5;

    for (int i = 0; i < 100; i++) {
        double x_next = phi(xIt);
        double d = abs(x_next - xIt);

        xnIter.push_back(xIt);
        xnext_iter.push_back(x_next);
        diffIter.push_back(d);

        xIt = x_next;
        if (d < eps) break;
    }

    cout << "N     xn          xn+1        |xn+1-xn|" << endl;
    for (int i = 0; i < xnIter.size(); i++) {
        cout << i << "     " << xnIter[i] << "    " << xnext_iter[i] << "    " << diffIter[i] << endl;
    }
    cout << "Корень: " << xIt << endl;

    return 0;
}

File: laba4/task1.go
package main

import (
	"fmt"
	"math"
)

// Исходная функция f(x)
func f(x float64) float64 {
	return math.Cos(x+0.5) - x - 2
}

// Производная f(x) для метода Ньютона
func df(x float64) float64 {
	return -math.Sin(x+0.5) - 1
}

// Функция x = phi(x) для метода итераций
func phi(x float64) float64 {
	return math.Cos(x+0.5) - 2
}

func main() {
	eps := 0.0001

	// --- Метод половинного деления ---
	fmt.Println("--- Метод половинного деления ---")
	a, b := -2.0, -1.0
	
	fmt.Printf("%-3s %-10s %-10s %-10s\n", "N", "an", "bn", "bn-an")
	
	iteration := 0
	for math.Abs(b-a) > eps {
		diff := math.Abs(b - a)
		fmt.Printf("%-3d %-10.5f %-10.5f %-10.5f\n", iteration, a, b, diff)
		
		c := (a + b) / 2.0
		if f(a)*f(c) < 0 {
			b = c
		} else {
			a = c
		}
		iteration++
	}
	fmt.Printf("Корень: %.5f\n\n", (a+b)/2.0)

	// --- Метод Ньютона ---
	fmt.Println("--- Метод Ньютона ---")
	xNewt := -1.5
	fmt.Printf("%-3s %-10s %-10s %-10s\n", "N", "xn", "xn+1", "|xn+1-xn|")
	
	for i := 0; i < 100; i++ {
		xNext := xNewt - f(xNewt)/df(xNewt)
		d := math.Abs(xNext - xNewt)
		
		fmt.Printf("%-3d %-10.5f %-10.5f %-10.5f\n", i, xNewt, xNext, d)
		
		xNewt = xNext
		if d < eps {
			break
		}
	}
	fmt.Printf("Корень: %.5f\n\n", xNewt)

	// --- Метод простых итераций ---
	fmt.Println("--- Метод простых итераций ---")
	xIt := -1.5
	fmt.Printf("%-3s %-10s %-10s %-10s\n", "N", "xn", "xn+1", "|xn+1-xn|")
	
	for i := 0; i < 100; i++ {
		xNext := phi(xIt)
		d := math.Abs(xNext - xIt)
		
		fmt.Printf("%-3d %-10.5f %-10.5f %-10.5f\n", i, xIt, xNext, d)
		
		xIt = xNext
		if d < eps {
			break
		}
	}
	fmt.Printf("Корень: %.5f\n", xIt)
}